SEMV 2.0: Patch doesn't need to start with a letter

tdavis opened the pull request:
It also allows the period (.) character due to v2.0 using it to delimit patch "pieces". Note that this commit does NOT address the following requirements:

- Identifiers MUST NOT be empty. 
- Numeric identifiers MUST NOT include leading zeroes.

When I get more time for C writin' I'll finish the rest of the 2.0 support...

theory commented:
I think you're right about it not requiring that it start with a letter; that requirement dates from [semver 1.0.0-beta](http://semver.org/spec/v1.0.0-beta.html), against which this extension was written. However, the period is *not* allows in [semver 1.0.0](http://semver.org/spec/v1.0.0.html), which is the version this extension supports. So I think we drop the first letter requirement, but not add the period support until [semver 2.0.0](http://semver.org/spec/v2.0.0.html) support is added.

tdavis commented:
Okay, I have completed SEMV 2.0.0 support and updated associated docs/tests appropriately. The only breaking change I made was to remove support for the transitional ``X.Y.Zpre`` style except where ``to_semver()`` is used explicitly.

theory commented:
Wow, nice, thanks. A couple of questions, points:

* Some stuff got deleted that didn't need to be, like the `gmake` bits of the README, and some whitespace was changed or removed. Would you mind restoring those bits, where possible? Spaces are preferred over tabs.
* I think we need a whole set of new tests for the expanded 2.0 versions. Like lots of examples with multiple dots after the dash, e.g., `1.4.2-2.12423.123.123`. Tests for proper sorting of such values would also be valuable.
* `<ctype.h>` is `#include`d twice, and there's a typo in one of the links in the README, pointing to `v22.0.0.` instead of `v2.0.0.`.
* What's `clean_junk()` for?

Thanks, this looks awesome!

tdavis commented:
David E. Wheeler writes:

> Wow, nice, thanks. A couple of questions, points:
>
> * Some stuff got deleted that didn't need to be, like the `gmake` bits of the
> README, and some whitespace was changed or removed. Would you mind restoring
> those bits, where possible? Spaces are preferred over tabs.

I can put the gmake stuff back, just seemed superfluous given the previous
block. I'll switch everything to four spaces, I just happened to format the
whole buffer as two but it makes no difference to me.

> * I think we need a whole set of new tests for the expanded 2.0 versions. Like
> lots of examples with multiple dots after the dash, e.g.,
> `1.4.2-2.12423.123.123`. Tests for proper sorting of such values would also be
> valuable.

Ok.

> * `<ctype.h>` is `#include`d twice, and there's a typo in one of the links in
> the README, pointing to `v22.0.0.` instead of `v2.0.0.`.

Will fix.

> * What's `clean_junk()` for?

It's used to remove periods and plus signs for comparison purposes, but looking
at it now it needs to be revised because everything after + should be thrown
out. I'll just adjust it based on the results of adding 2.0 comp tests.

>
> Thanks, this looks awesome!
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/theory/pg-semver/pull/13#issuecomment-116851813

tdavis commented:
I believe I addressed all the issues you mentioned. Let me know!

theory commented:
Couple of warnings from Clang:

    src/semver.c:91:10: warning: incompatible pointer to integer conversion initializing 'char' with an expression of
          type 'void *' [-Wint-conversion]
        char next = NULL;
             ^      ~~~~
    src/semver.c:313:28: warning: unsequenced modification and access to 'j' [-Wunsequenced]
                copy[j] = str[j++];
                     ~         ^
    2 warnings generated.

theory commented:
Also, I get test failures:

    *** /Users/david/dev/pgxn/pg-semver/test/expected/base.out	2015-07-02 13:38:21.000000000 -0700
    --- /Users/david/dev/pgxn/pg-semver/results/base.out	2015-07-02 13:40:35.000000000 -0700
    ***************
    *** 63,70 ****
      ok 61 - v1.2.3 should not equal v1.2.3-b
      ok 62 - semver(1.2.3-a, 1.2.3-b) should <> 0
      ok 63 - v1.2.3-a should not equal v1.2.3-b
    ! ok 64 - semver(1.2.3-aaaaaaa1, 1.2.3-aaaaaaa2) should <> 0
    ! ok 65 - v1.2.3-aaaaaaa1 should not equal v1.2.3-aaaaaaa2
      ok 66 - semver(2.2.2, 1.1.1) should > 0
      ok 67 - semver(1.1.1, 2.2.2) should < 0
      ok 68 - v2.2.2 should be > v1.1.1
    --- 63,72 ----
      ok 61 - v1.2.3 should not equal v1.2.3-b
      ok 62 - semver(1.2.3-a, 1.2.3-b) should <> 0
      ok 63 - v1.2.3-a should not equal v1.2.3-b
    ! not ok 64 - semver(1.2.3-aaaaaaa1, 1.2.3-aaaaaaa2) should <> 0
    ! # Failed test 64: "semver(1.2.3-aaaaaaa1, 1.2.3-aaaaaaa2) should <> 0"
    ! not ok 65 - v1.2.3-aaaaaaa1 should not equal v1.2.3-aaaaaaa2
    ! # Failed test 65: "v1.2.3-aaaaaaa1 should not equal v1.2.3-aaaaaaa2"
      ok 66 - semver(2.2.2, 1.1.1) should > 0
      ok 67 - semver(1.1.1, 2.2.2) should < 0
      ok 68 - v2.2.2 should be > v1.1.1
    ***************
    *** 101,111 ****
      ok 99 - v2.2.2-c should be >= v2.2.2-b
      ok 100 - v2.2.2-b should be < v2.2.2-c
      ok 101 - v2.2.2-b should be <= v2.2.2-c
    ! ok 102 - semver(2.2.2-rc-2, 2.2.2-RC-1) should > 0
    ! ok 103 - semver(2.2.2-RC-1, 2.2.2-rc-2) should < 0
    ! ok 104 - v2.2.2-rc-2 should be > v2.2.2-RC-1
      ok 105 - v2.2.2-rc-2 should be >= v2.2.2-RC-1
    ! ok 106 - v2.2.2-RC-1 should be < v2.2.2-rc-2
      ok 107 - v2.2.2-RC-1 should be <= v2.2.2-rc-2
      ok 108 - semver(0.9.10, 0.9.9) should > 0
      ok 109 - semver(0.9.9, 0.9.10) should < 0
    --- 103,117 ----
      ok 99 - v2.2.2-c should be >= v2.2.2-b
      ok 100 - v2.2.2-b should be < v2.2.2-c
      ok 101 - v2.2.2-b should be <= v2.2.2-c
    ! not ok 102 - semver(2.2.2-rc-2, 2.2.2-RC-1) should > 0
    ! # Failed test 102: "semver(2.2.2-rc-2, 2.2.2-RC-1) should > 0"
    ! not ok 103 - semver(2.2.2-RC-1, 2.2.2-rc-2) should < 0
    ! # Failed test 103: "semver(2.2.2-RC-1, 2.2.2-rc-2) should < 0"
    ! not ok 104 - v2.2.2-rc-2 should be > v2.2.2-RC-1
    ! # Failed test 104: "v2.2.2-rc-2 should be > v2.2.2-RC-1"
      ok 105 - v2.2.2-rc-2 should be >= v2.2.2-RC-1
    ! not ok 106 - v2.2.2-RC-1 should be < v2.2.2-rc-2
    ! # Failed test 106: "v2.2.2-RC-1 should be < v2.2.2-rc-2"
      ok 107 - v2.2.2-RC-1 should be <= v2.2.2-rc-2
      ok 108 - semver(0.9.10, 0.9.9) should > 0
      ok 109 - semver(0.9.9, 0.9.10) should < 0
    ***************
    *** 194,200 ****
      ok 192 - "1.0.0+1-1" is not a valid 2.0.0 semver
      ok 193 - "1.0.0-1...." is not a valid 2.0.0 semver
      ok 194 - ORDER BY semver (2.0.0) USING < should work
    ! ok 195 - ORDER BY semver (2.0.0) USING > should work
      ok 196 - semver(1.0.0-1+1, 1.0.0-1+5) should = 0
      ok 197 - v1.0.0-1+1 should = v1.0.0-1+5
      ok 198 - v1.0.0-1+1 should be <= v1.0.0-1+5
    --- 200,210 ----
      ok 192 - "1.0.0+1-1" is not a valid 2.0.0 semver
      ok 193 - "1.0.0-1...." is not a valid 2.0.0 semver
      ok 194 - ORDER BY semver (2.0.0) USING < should work
    ! not ok 195 - ORDER BY semver (2.0.0) USING > should work
    ! # Failed test 195: "ORDER BY semver (2.0.0) USING > should work"
    ! #     Results differ beginning at row 4:
    ! #         have: (0.9.9-a1.1+234)
    ! #         want: (0.9.9-a1.2)
      ok 196 - semver(1.0.0-1+1, 1.0.0-1+5) should = 0
      ok 197 - v1.0.0-1+1 should = v1.0.0-1+5
      ok 198 - v1.0.0-1+1 should be <= v1.0.0-1+5
    ***************
    *** 205,207 ****
    --- 215,218 ----
      ok 203 - v1.0.0-1.1+1 should be >= v1.0.0-1.1+5
      ok 204 - Should correctly represent "0.5.0-release1" as "0.5.0-release1"
      ok 205 - Should correctly represent "0.5.0release1" as "0.5.0-release1"
    + # Looks like you failed 7 tests of 205

    ======================================================================

tdavis commented:
I fixed the clang errors, but I can't reproduce the test failures. I have been developing by changing the source then running ``sudo make install && make installcheck``. The tests pass and ``results/base.out`` reflects that with all "ok". I'm not certain why it would make a difference, but I'm using v9.4.2 of Postgres, what are you using?

theory commented:
The tests pass for me now, yay!

I pulled your changes into the [semver2 branch](/theory/pg-semver/tree/semver2). I added some more tests to as a sanity check, and in reviewing the spec, it says, "Numeric identifiers MUST NOT include leading zeroes." So I added some a test with a leading zero and it failed. I think in strict (normal) mode, a leading zero should be disallowed, but `to_semver()` can simply remove them. This is how it works for the `x.y.z` part of the expression. Can you fix that issue?

Otherwise I think we're good. This is great, and thanks!

tdavis commented:
I have merged with your branch and added fixes so all additional tests now pass. I also fixed a regression I discovered regarding pre-release values and lax parsing, which I also added a couple tests for.

theory commented:
Okay, which commits do I need to cherry-pick, then?

tdavis commented:
Sorry, added one more fix! You shouldn't need to cherry-pick anything, I merged in changes from your semver2 branch.

theory commented:
Looks great, thank you!

I don't think this changes the representation stored by the database, right? Nor did we change any SQL. So it should be perfectly compatible other than the removal of the transitional `X.Y.Zpre` style you mentioned, correct?

tdavis commented:
Yep! I didn't change anything aside from the logic of parsing. Ordering
just ignores the newer metadata (+) stuff, so it should be transparent to
the user.

On Wed, Jul 8, 2015, 6:20 PM David E. Wheeler <notifications@github.com>
wrote:

> Looks great, thank you!
>
> I don't think this changes the representation stored by the database,
> right? Nor did we change any SQL. So it should be perfectly compatible
> other than the removal of the transitional X.Y.Zpre style you mentioned,
> correct?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/theory/pg-semver/pull/13#issuecomment-119747986>.
>

theory commented:
Awesome, thank you!

theory commented:
I'm thinking it might be time for a release. Have you been using this code, @tdavis?

tdavis commented:
Yep, never had any issues with it.

theory commented:
Thanks, [released](http://pgxn.org/dist/semver/)!

I [added you to the list of authors](https://github.com/theory/pg-semver/commit/5ff2bdcc70194b35000f85e65a2c82c9ca94b612#diff-3a79cbc53237b76800aeadfa348c0869L191), BTW. Hope you don't mind. Thanks for your contribution!

tdavis commented:
Yay! I don't mind at all, thanks for the attribution!
